Implementing Client-side UX improvements

axelvyrn opened the pull request:
Closes #43 

Changes in this PR:
- Introduces memoization and deduplication for event subscriptions in `NostrRelayManager`, including in-memory event caching with TTL and active subscription deduplication. 
- Adds relay backoff logic to handle rate-limited or unhealthy relays, improving reliability and efficiency. 
- Updates `useEngagementMetrics` to rely on relay manager for deduplication and clarifies subscription tracking.

axelvyrn committed:
Add deduplication and relay backoff to relay manager

Introduces memoization and deduplication for event subscriptions in NostrRelayManager, including in-memory event caching with TTL and active subscription deduplication. Adds relay backoff logic to handle rate-limited or unhealthy relays, improving reliability and efficiency. Updates useEngagementMetrics to rely on relay manager for deduplication and clarifies subscription tracking.

axelvyrn commented:
Not yet ready for review, since I haven't fully checked whether all features are functioning or not.
I'll be attaching screenshots together.

But so far, what I'm sure of is that the NWC settings remember the relays and handles errors properly
<img width="962" height="489" alt="image" src="https://github.com/user-attachments/assets/68f0ebea-73aa-493f-b137-dc41cb07994c" />

DoktorShift commented:
@axelvyrn  while working on it, please don’t test only with “empty” Nostr Identitys. Also use active account with (in best case) plenty of notes, content, zapgoals, zaps, follow pack members, and engagements to ensure the code handles larger data loads in best practice. The goal is twofold:

- Improve the technical side and code quality.
- Provide clear user feedback (e.g. user friendly loading states) if longer loading times are unavoidable.

axelvyrn commented:
Yes, I have created a nostr identity from my side. I have checked relays and chats so far. I'll be checking and providing status of the rest soon.

axelvyrn committed:
Improve large dataset handling and loading progress UI

Adds progressive loading indicators and batching for profile fetching to enhance performance with large datasets. Increases event subscription limits in useNostrLongForm.js and useNostrNotes.js to support users with many notes or long-form posts. Updates the UI to show progress bars, performance info, and summary banners for large data loads, improving user feedback and experience during heavy operations.

axelvyrn commented:
Okay so keeping in mind, smooth transitions, first loading time is drastically reduced from ~30s to ~5-10s due to progressive loading. Content, notes and chats are working as expected. Follow packs and suggestions are working fine too.
<img width="493" height="641" alt="image" src="https://github.com/user-attachments/assets/a24adddf-daf5-47d8-98cd-07c28ca20e4f" />
<img width="1043" height="463" alt="image" src="https://github.com/user-attachments/assets/7bc8ff1e-f2c5-480b-8b73-a60ebcb90d80" />
<img width="999" height="333" alt="image" src="https://github.com/user-attachments/assets/c68dadcb-554a-4d94-b9be-e3ebc6fe6897" />
<img width="1262" height="617" alt="image" src="https://github.com/user-attachments/assets/ed237633-678f-40d8-87ec-cf40833ed83f" />


Zaps are working fine too, tested through script tests.

## New Logging Statements and When They Appear (App.vue):

**1. Profile Fetching Logs (Lines 302, 318):**
`When they appear:` Only when there are zaps with sender profiles that haven't been fetched yet. This happens when:
You have zaps from users whose profiles aren't cached
The profile store doesn't already have their profile data
**2. Performance Logs (Line 429):**
`When it appears: `Only when you have more than 100 zaps in your dataset. This is the threshold I set to avoid spam for small datasets.
**3. Data Refresh Logs (Lines 411, 441):**
`When they appear: `Every time data is refreshed (on connection, manual refresh, periodic refresh).

### Why You Might Not See Them:

- Small Dataset: If you have fewer than 100 zaps, the performance log won't show
- No New Profiles: If all zap senders' profiles are already cached, profile fetching logs won't show
- No Wallet Connection: If no wallet is connected, most logs won't appear
- Console Not Open: Make sure you have the browser's developer console open (F12) (yeah people forget)

### To Test the Logging:
- Open Browser Console
- Connect a wallet with some transaction history
- Look for these specific logs:
`"Refreshing zap data..."`
`"Zap data refreshed successfully: X zaps"`
If you have 100+ zaps: `"📊 Performance: Loaded X zaps in Yms"`
If there are new profiles: `"📊 Fetching X profiles in batches..."`

axelvyrn commented:
Max long forms and notes increased to 200 from 100 to improve user metrics. Logging and error handling are managed effectively in App.vue
Removed duplication of fallback avatar generation code in App.vue

Introduces memoization and deduplication for event subscriptions in NostrRelayManager, including in-memory event caching with TTL and active subscription deduplication. Adds relay backoff logic to handle rate-limited or unhealthy relays, improving reliability and efficiency. Updates useEngagementMetrics to rely on relay manager for deduplication and clarifies subscription tracking.

DoktorShift commented:
hey @axelvyrn,  

I’m testing your PR right now and I can feel a **performance boost** – great job on that!  

#### What I have to mention  

1. **Audience Tab**  
   - Needs a bit more attention → Followers, Follower Profiles, Follow Packs, and Suggestions are quite heavy loads.  
   - At the moment, the UI struggles: profile pictures, names, and general profile data are not displayed correctly.  

<img width="1911" height="912" alt="grafik" src="https://github.com/user-attachments/assets/9d421785-0eda-4cc1-a3ac-84b8136f4a30" />  

<img width="1342" height="804" alt="grafik" src="https://github.com/user-attachments/assets/28ef8f9f-c951-4df1-b309-057f1cba0a45" />  

<img width="1908" height="918" alt="grafik" src="https://github.com/user-attachments/assets/984ee21c-24ca-4c23-a0ee-697c1bc7f2e7" />  

I also noticed a discrepancy between the loaded and used amount of zaps i can not explain myself right now.  

<img width="1912" height="956" alt="grafik" src="https://github.com/user-attachments/assets/88491538-9e03-43b0-b1eb-f26a2ccc492d" />  

---


2. **ZapTracker Behavior**  
   - **Current:** On browser refresh (often seen user behaviour) , data only loads when components mount, means for example ZapFeed will be empty als long as the user does not mount the single components.  
   - **Expected:** On refresh (while logged in), **cached data should appear instantly** and a controlled refresh should be done in background.
 


---

Overall: this PR already brings a boost in performance and a lot of value to the project. At the same time, these two remaining areas are still important to refine - once addressed, they’ll really complete the picture and turn this PR into a very powerful update for all users. 🙂

axelvyrn commented:
On it

axelvyrn committed:
Centralize and share Nostr profile fetching logic

Refactored profile fetching and caching into a new shared utility (profileFetcher.js) and updated useAudience, useFollowLists, and useNostrChat to use this centralized logic. This reduces code duplication, ensures consistent profile caching across modules, and improves batch profile fetching for efficiency. Also includes minor normalization and cleanup in follow list and audience composables.

axelvyrn commented:
<img width="1031" height="406" alt="image" src="https://github.com/user-attachments/assets/30d3225a-c73b-466f-bcd1-a50a847ef01b" />
Random "user:c0123f" strings won't appear anymore. Looks like only the kind:3 follow lists was being fetched and not the kind:0 metadata

axelvyrn commented:
@DoktorShift well, I must inform you of something, this is something I can't help about
<img width="991" height="338" alt="image" src="https://github.com/user-attachments/assets/1d45a908-4a7d-4e19-852d-8b9e50490b9d" />
As you can see here, Liberty Gal has her username and pfp correct, after my profile fetching correction while the next one does not
That guy is
<img width="1306" height="620" alt="image" src="https://github.com/user-attachments/assets/efa6882b-5bb5-4352-bce4-f15ce0932a9f" />
But his info is not fetched because he doesn't have a kind:0 published or updated. In fact,
Many npubs are just raw keys with no metadata.
If a pubkey never broadcast a kind-0 event, you’ll always see fallback.
So, unfortunately the client should gracefully fallback to shortPubkey, which is the "user:77dc92e2"
Kind-0 is not “globally” stored — only relays where the user published it. But zap tracker does connect to the most popular ones, so that's really depending on the respective users.

axelvyrn committed:
Refactor to batch fetch profiles for efficiency

Replaced individual fetchProfile calls with batchFetchProfiles in useFollowLists.js to improve efficiency when fetching profiles for members and new follows.

axelvyrn committed:
Add immediate cache load and background refresh to composables

Each of useAudience, useEngagementMetrics, and useFollowLists now loads cached data from localStorage on startup to improve UI responsiveness after page refresh. Additionally, background refresh logic is added to update cached data once the relay manager is initialized, ensuring data stays up-to-date without blocking the main thread.

axelvyrn commented:
Well, I was surprised myself by how fast the app now loads, almost instantly. Feel free to check out.
Only the zap data bug is left now.

axelvyrn committed:
Count only zaps with eventId in status message

Updated the status message to display the count of zaps that have an eventId, providing a more accurate representation of loaded zaps.

axelvyrn commented:
I found the mismatch: the green “Large dataset loaded successfully” banner used enhancedCombinedZapData.length (all items), while the sidebar and analytics use only NIP-57 zaps with eventId.
I updated the banner to count only NIP-57 zaps, so it now displays the same total as the left modal.
All done @DoktorShift

DoktorShift commented:
I’ll review your changes during the day - looks promising so far 🙂 

If nothing unexpected pops up, we should be able to merge this afternoon or tomorrow and release your funds 🎉  

In the meantime, maybe you can tackle this small UI issue so we can settle everything nicely:  

---

### Green large dataset banner (success box) – `App.vue`  

**Current issue**  
- Banner stays visible permanently, user cannot close it.  

**Expected behavior**  
- Add a small **“X” close button** (top-right or right side) so the user can dismiss it manually.  
- Auto-hide the banner **after 5 seconds** on every load.  
- Use a **smooth fade-out / slide-away animation** when it disappears.  

---

<img width="1880" height="908" alt="grafik" src="https://github.com/user-attachments/assets/5557b001-0af7-428a-9feb-f56da6183c52" />

axelvyrn committed:
Add auto-hide and dismiss for large dataset banner

Introduces UI state and logic to automatically show and hide the large dataset banner after 5 seconds, and adds a manual dismiss button. This improves user experience by preventing persistent banners when viewing large datasets.

axelvyrn commented:
That should do, I couldn't provide the screenshot I'm sorry because the large dataset is not something I still have in the demo accounts I am using. Although I have used that same logic elsewhere in other projects too so I'm almost sure it would work. If not, please do let me know.

pratik227 committed:
Merge branch 'dev' into dev

DoktorShift commented:
@axelvyrn 

Thanks for your work on this task. 
With your PR ZapTracker got some good performance upgrade.

axelvyrn commented:
Glad to help!
